plugins/instruction_mix.c

```diff
@@ -0,0 +1,262 @@
+  counters->store = 0;
+  counters->branch = 0;
+#ifdef COUNT_PRFM
+  counters->prefetch = 0;
+#endif
+}
+
+int instruction_count_post_thread_handler(mambo_context *ctx) {
+  // On thread exit, the counters are added to the global counters
+  struct instructions *counters = mambo_get_thread_plugin_data(ctx);
+
+  fprintf(stderr, "Thread: %d\n", mambo_get_thread_id(ctx));
+
+  // Prints thread private counters
+  print_counters(counters); // comment this out if not needed
+
+  atomic_increment_u64(&global_counters.integer, counters->integer);
+  atomic_increment_u64(&global_counters.floating, counters->floating);
+  atomic_increment_u64(&global_counters.load, counters->load);
+  atomic_increment_u64(&global_counters.store, counters->store);
+  atomic_increment_u64(&global_counters.branch, counters->branch);
+#ifdef COUNT_PRFM
+  atomic_increment_u64(&global_counters.prefetch, counters->prefetch);
+#endif
+  mambo_free(ctx, counters);
+}
+
+int instruction_count_pre_inst_handler(mambo_context *ctx) {
+  struct instructions *counters = mambo_get_thread_plugin_data(ctx);
+  uint64_t *inst_counter = NULL;
+
+#ifdef __aarch64__
+  // Variables are used for decoding the fields of intructions
+	uint32_t sf, rm, opcode, rn, rd;
+
+  // Selects the appropriate counter according to the instruction type
+  switch (ctx->code.inst) {
+  // Branches, Exception Generating and System instructions Category
+      // * Branches instructions
+  case A64_B_BL:
+  case A64_B_COND:
+  case A64_CBZ_CBNZ:
+  case A64_TBZ_TBNZ:
+  case A64_BR:
+  case A64_BLR:
+  case A64_RET:
+    inst_counter = &counters->branch;
+    break;
+
+      // * Exception Generating
+  case A64_SVC:
+  case A64_HVC:
+  case A64_BRK:
+    break;
+
+      // * System instructions
+  case A64_SYS:
+  case A64_MRS_MSR_REG:
+  case A64_HINT:
+  case A64_DSB:
+  case A64_DMB:
+  case A64_ISB:
+  case A64_CLREX:
+    break;
+
+  // Data Processing -- Immediate
+  case A64_ADD_SUB_IMMED:
+  case A64_LOGICAL_IMMED:
+    inst_counter = &counters->integer;
+    break;
+
+  case A64_BFM:
+  case A64_ADR:
+  case A64_EXTR:
+  case A64_MOV_WIDE:
+    break;
+
+  // Loads and Stores
+  case A64_LDR_LIT:
+  case A64_LDX_STX:
+  case A64_LDP_STP:
+  case A64_LDR_STR_IMMED:
+  case A64_LDR_STR_REG:
+  case A64_LDR_STR_UNSIGNED_IMMED:
+  case A64_LDX_STX_MULTIPLE:
+  case A64_LDX_STX_MULTIPLE_POST:
+  case A64_LDX_STX_SINGLE:
+  case A64_LDX_STX_SINGLE_POST:
+    if (mambo_is_load(ctx)) {
+      inst_counter = &counters->load;
+    } else if (mambo_is_store(ctx)) {
+      inst_counter = &counters->store;
+    } else {
+#ifdef COUNT_PRFM
+      inst_counter = &counters->prefetch;
+#endif
+    }
+    break;
+
+  // Data Processing -- Register
+  case A64_ADD_SUB_EXT_REG:
+  case A64_ADD_SUB_SHIFT_REG:
+  case A64_ADC_SBC:
+    inst_counter = &counters->integer;
+    break;
+  case A64_DATA_PROC_REG1:
+  case A64_CCMP_CCMN_IMMED:
+  case A64_CCMP_CCMN_REG:
+  case A64_COND_SELECT:
+  case A64_LOGICAL_REG:
```

Comment:
I'd classify all of these up to and including A64_DATA_PROC_REG3 as integer instructions.

Reply:
Done